packages/client/package.json

```diff
@@ -3,8 +3,14 @@
   "version": "0.0.0",
   "scripts": {
     "dev": "webpack-dev-server",
-    "build": "NODE_ENV=production webpack -p; cp _redirects dist/_redirects"
+    "devWebpack": "webpack -p",
+
+    "build-mac": "NODE_ENV=production webpack -p; cp _redirects dist/_redirects",
+    "build-win": "set NODE_ENV=development && npm run devWebpack && copy .\\_redirects .\\dist",
```

Comment:
Can we use [`cross-env`](https://www.npmjs.com/package/cross-env) for this? I think it will be a much better solution than seperate build commmands.

As for the copy command, we can also use the [`copy-webpack-plugin`](https://github.com/webpack-contrib/copy-webpack-plugin) instead. That way we don't need to do it from the commands, and we can just have a single `yarn build`

Reply:
I will look into cross-env

Reply:
So I have done some research on **cross-env** and **command prompt** behaviors.  
This is what I have found.

## **cross-env issue**
It works on my window machine.  I cannot test it on Mac or Linux.  
I will change my laptop OS to Linux since Windows 10 don't work well on it.

## **Original Script Issue**
The biggest dilemma that I am having is the usage of **&&** and **;** within command line, because I simply cannot get this line of the script to work inside the client package.json within Window.  To understand why, I have isolated the test to see which part of the script is giving me problems.

```
"NODE_ENV=production webpack -p; cp _redirects dist/_redirects"
```

After doing a lot of testing, these are the issues that I have run into:
- Difference between **cp** and **copy**
- How path **"/"** is used
- Difference between **&&** and **;**



## **Difference between cp and copy**
This is a strange issue because the cp command doesn't work with npm or yarn.
The cp command only works when it is directly typed into the command line.  I am taking this as a windows specific issue.

When executing the cp command, it will generate this error ```ERROR in multi ./_redirects dist/_redirects```

To run more test, I have isolated the command to either cp or copy.
**The result**: cp-filed failed, copy-file made the copy.
```
"cp-file": "cp sample.txt dist",
"copy-file": "copy sample.txt dist",
```


The **solution** I have is to build a window specific script that uses the copy command.
```
"NODE_ENV=production webpack -p; copy _redirects dist/_redirects",
```
This leads to the 2nd issue aside from cp command.

## **How path "/" is used**
Similar to cp and copy issues, there is a difference between **"/"** and **"```\\```"** on windows.

**Sample Isolated test:**
```
"copy-frwd": "copy sample.txt dist/",
"copy-bkwd": "copy sample.txt dist\\",
```
**Result:** copy-frwd failed, and copy-bkwd is able to complete the copy.
```
"NODE_ENV=production webpack -p; copy _redirects dist\\_redirects",
```

## Difference between **&&** and **;**
Again, similar to copy, and / path.  ; only seem to work on command prompt directly and not within npm or yarn environment.


I have research into the difference between && and ;
**&& is a logical operator**, while **; is for sequencing**.

What I learned is...
```
cmd1 && cmd2   //cmd2 only executes if cmd1 returns a success code
```
_Another words, cmd2 depends on cmd1 to execute properly._

```
cmd1; cmd2   //cmd1 and cmd2 will executes regardless if one fails
```
_If cmd2 depends on cmd1, cmd2 will run regardless if cmd1 fails._

To be sure if I should keep the ; or &&, I ran the following test.
Create a file with text inside and open it all within the same command.
**Sample Test Script**
```
"create-file": "echo 'hello world' > sample.txt",
"open-file": "notepad sample.txt",

"file-semi":"npm run create-file; npm run open-file",
"file-and": "npm run create-file && npm run open-file"
```
**Result:** **file-semi** failed, while **fail-and** successfully execute both commands in sequence.
**Error** that came up is this...

```
$ npm run create-file; npm run open-file
npm ERR! missing script: create-file;
```

### Conclusion
After many hours of tinkering with the script.  I do not have any solution to counter the ; colon issue to have only 1 script.

I separate the build command in the client between mac build and win build, while using a build.js to selectively execute which one to use based on the OS system.

**The Final Script**
```
"build-mac": "cross-env NODE_ENV=production webpack -p; cp _redirects dist/_redirects",
"build-win": "cross-env NODE_ENV=development webpack -p && copy .\\_redirects .\\dist",
"build": "node build.js"
```
Now I see why you don't like Windows. 😄